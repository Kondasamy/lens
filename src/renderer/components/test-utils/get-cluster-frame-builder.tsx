/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import type { DiContainer } from "@ogre-tools/injectable";
import type { LensRendererExtension } from "../../../extensions/lens-renderer-extension";
import rendererExtensionsInjectable from "../../../extensions/renderer-extensions.injectable";
import currentlyInClusterFrameInjectable from "../../routes/currently-in-cluster-frame.injectable";
import { extensionRegistratorInjectionToken } from "../../../extensions/extension-loader/extension-registrator-injection-token";
import { computed } from "mobx";
import { renderFor } from "./renderFor";
import observableHistoryInjectable from "../../navigation/observable-history.injectable";
import currentRouteComponentInjectable from "../../routes/current-route-component.injectable";
import { Sidebar } from "../layout/sidebar";
import React from "react";
import { Router } from "react-router";
import { Observer } from "mobx-react";
import subscribeStoresInjectable from "../../kube-watch-api/subscribe-stores.injectable";
import allowedResourcesInjectable from "../../../common/cluster-store/allowed-resources.injectable";
import type { RenderResult } from "@testing-library/react";

interface Options {
  di: DiContainer;
  extensions: LensRendererExtension[];
}

type Callback = () => void;

export interface ClusterFrameBuilder {
  beforeSetups: (callback: Callback) => ClusterFrameBuilder;
  beforeRender: (callback: Callback) => ClusterFrameBuilder;
  render: () => Promise<RenderResult>;
}

export const getClusterFrameBuilder = ({ di, extensions = [] }: Options) => {
  const beforeSetupsCallbacks: Callback[] = [];
  const beforeRenderCallbacks: Callback[] = [];

  di.override(subscribeStoresInjectable, () => () => () => {});

  di.override(rendererExtensionsInjectable, () => computed(() => extensions));

  di.override(currentlyInClusterFrameInjectable, () => true);

  di.override(allowedResourcesInjectable, () =>
    computed(() => new Set<string>()),
  );

  const builder: ClusterFrameBuilder = {
    beforeSetups(callback: () => void) {
      beforeSetupsCallbacks.push(callback);

      return builder;
    },

    beforeRender(callback: () => void) {
      beforeRenderCallbacks.push(callback);

      return builder;
    },

    async render() {
      beforeSetupsCallbacks.forEach(callback => callback());

      await di.runSetups();

      const extensionRegistrators = di.injectMany(
        extensionRegistratorInjectionToken,
      );

      await Promise.all(
        extensions.flatMap((extension) =>
          extensionRegistrators.map((registrator) => registrator(extension, 1)),
        ),
      );

      const render = renderFor(di);

      const history = di.inject(observableHistoryInjectable);
      const currentRouteComponent = di.inject(currentRouteComponentInjectable);

      beforeRenderCallbacks.forEach(callback => callback());

      return render(
        <Router history={history}>
          <Sidebar />

          <Observer>
            {() => {
              const Component = currentRouteComponent.get();

              if (!Component) {
                return null;
              }

              return <Component />;
            }}
          </Observer>
        </Router>,
      );
    },
  };

  return builder;
};
