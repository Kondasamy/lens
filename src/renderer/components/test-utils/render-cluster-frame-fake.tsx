/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import type { DiContainer } from "@ogre-tools/injectable";
import type { LensRendererExtension } from "../../../extensions/lens-renderer-extension";
import rendererExtensionsInjectable from "../../../extensions/renderer-extensions.injectable";
import currentlyInClusterFrameInjectable from "../../routes/currently-in-cluster-frame.injectable";
import type { Cluster } from "../../../common/cluster/cluster";
import hostedClusterInjectable from "../../../common/cluster-store/hosted-cluster.injectable";
import { extensionRegistratorInjectionToken } from "../../../extensions/extension-loader/extension-registrator-injection-token";
import { computed } from "mobx";
import { renderFor } from "./renderFor";
import observableHistoryInjectable from "../../navigation/observable-history.injectable";
import currentRouteComponentInjectable from "../../routes/current-route-component.injectable";
import { Sidebar } from "../layout/sidebar";
import React from "react";
import { Router } from "react-router";
import { Observer } from "mobx-react";
import subscribeStoresInjectable from "../../kube-watch-api/subscribe-stores.injectable";

interface Options {
  di: DiContainer;
  extensions: LensRendererExtension[];
}

export const renderClusterFrameFake = async ({ di, extensions = [] }: Options) => {
  di.override(subscribeStoresInjectable, () => () => () => {});

  di.override(rendererExtensionsInjectable, () => computed(() => extensions));

  di.override(currentlyInClusterFrameInjectable, () => true);

  const clusterStub = { allowedResources: [] } as Cluster;

  di.override(hostedClusterInjectable, () => clusterStub);

  await di.runSetups();

  const extensionRegistrators = di.injectMany(
    extensionRegistratorInjectionToken,
  );

  await Promise.all(
    extensions.flatMap((extension) =>
      extensionRegistrators.map((registrator) => registrator(extension, 1)),
    ),
  );

  const render = renderFor(di);

  const history = di.inject(observableHistoryInjectable);
  const currentRouteComponent = di.inject(currentRouteComponentInjectable);

  return () => render(
    <Router history={history}>
      <Sidebar />

      <Observer>
        {() => {
          const Component = currentRouteComponent.get();

          if (!Component) {
            return null;
          }

          return <Component />;
        }}
      </Observer>
    </Router>,
  );
};
