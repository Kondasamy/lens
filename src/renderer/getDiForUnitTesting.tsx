/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import glob from "glob";
import {
  join,
  memoize,
  split,
  noop,
  slice,
  filter,
  last,
  map,
  tap,
} from "lodash/fp";
import { createContainer } from "@ogre-tools/injectable";
import { setLegacyGlobalDiForExtensionApi } from "../extensions/as-legacy-globals-for-extension-api/legacy-global-di-for-extension-api";
import getValueFromRegisteredChannelInjectable from "./app-paths/get-value-from-registered-channel/get-value-from-registered-channel.injectable";
import loggerInjectable from "../common/logger.injectable";
import fsInjectable from "../common/fs/fs.injectable";
import type fse from "fs-extra";
import { pipeline } from "@ogre-tools/fp";

export const getDiForUnitTesting = (
  { doGeneralOverrides } = { doGeneralOverrides: false },
) => {
  const di = createContainer();

  setLegacyGlobalDiForExtensionApi(di);

  for (const filePath of getInjectableFilePaths()) {
    const injectableInstance = require(filePath).default;

    di.register({
      ...injectableInstance,
      aliases: [injectableInstance, ...(injectableInstance.aliases || [])],
    });
  }

  di.preventSideEffects();

  if (doGeneralOverrides) {
    di.override(getValueFromRegisteredChannelInjectable, () => () => undefined);

    di.override(fsInjectable, () => getFsFake());

    di.override(loggerInjectable, () => ({
      warn: noop,
      debug: noop,
      log: noop,
      error: (...args: any) => console.error(...args),
      info: noop,
    }));
  }

  return di;
};

const getInjectableFilePaths = memoize(() => [
  ...glob.sync("./**/*.injectable.{ts,tsx}", { cwd: __dirname }),
  ...glob.sync("../common/**/*.injectable.{ts,tsx}", { cwd: __dirname }),
  ...glob.sync("../extensions/**/*.injectable.{ts,tsx}", { cwd: __dirname }),
]);

const getFsFake = () => {
  const state = new Map();

  const readFile = (filePath: string) => {
    const fileContent = state.get(filePath);

    if (!fileContent) {
      const existingFilePaths = [...state.keys()].join('", "');

      throw new Error(
        `Tried to access file ${filePath} which does not exist. Existing file paths are: "${existingFilePaths}"`,
      );
    }

    return Promise.resolve(Buffer.from(fileContent));
  };

  return {
    ensureDir: () => Promise.resolve(),

    writeJson: (filePath: string, contents: object) => {
      state.set(filePath, JSON.stringify(contents));

      return Promise.resolve();
    },

    readJson: async (filePath: string) => {
      const fileContent = await readFile(filePath);

      return JSON.parse(fileContent.toString());
    },

    readFile,

    readdir: (directoryPath: string) => {
      const fileNames = pipeline(
        [...state.keys()],

        filter((x) =>
          pipeline(
            x,
            split("/"),
            slice(0, -1),
            join("/"),
          ),
        ),

        tap(console.log),

        map((x) => pipeline(x, split("/"), last)),
      );

      return Promise.resolve(fileNames);
    },
  } as unknown as typeof fse;
};
