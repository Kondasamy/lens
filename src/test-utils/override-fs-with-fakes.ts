/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import type { DiContainer } from "@ogre-tools/injectable";
import readFileInjectable from "../common/fs/read-file.injectable";
import writeJsonFileInjectable from "../common/fs/write-json-file.injectable";
import readJsonFileInjectable from "../common/fs/read-json-file.injectable";
import pathExistsInjectable from "../common/fs/path-exists.injectable";

export const overrideFsWithFakes = (di: DiContainer) => {
  const state = new Map();

  const readFile = readFileFor(state);

  const overrides = [
    {
      injectableToOverride: readFileInjectable,

      instantiateFake: () => readFile,
    },

    {
      injectableToOverride: writeJsonFileInjectable,

      instantiateFake: () => (filePath: string, contents: object) => {
        state.set(filePath, JSON.stringify(contents));

        return Promise.resolve();
      },
    },

    {
      injectableToOverride: readJsonFileInjectable,

      instantiateFake: () => async (filePath: string) => {
        const fileContent = await readFile(filePath);

        return JSON.parse(fileContent.toString());
      },
    },

    {
      injectableToOverride: pathExistsInjectable,

      instantiateFake: () => async (filePath: string) =>
        Promise.resolve(state.has(filePath)),
    },
  ];

  overrides.forEach(({ injectableToOverride, instantiateFake }) => {
    di.override(injectableToOverride, instantiateFake);
  });
};

const readFileFor = (state: Map<string, Buffer>) => (filePath: string) => {
  const fileContent = state.get(filePath);

  if (!fileContent) {
    const existingFilePaths = [...state.keys()].join('", "');

    throw new Error(
      `Tried to access file ${filePath} which does not exist. Existing file paths are: "${existingFilePaths}"`,
    );
  }

  return Promise.resolve(Buffer.from(fileContent));
};
