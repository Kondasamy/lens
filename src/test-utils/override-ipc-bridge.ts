/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import type { DiContainer } from "@ogre-tools/injectable";
import getValueFromChannelInjectable from "../renderer/channel/get-value-from-channel.injectable";
import registerChannelInjectable from "../main/app-paths/register-channel/register-channel.injectable";
import asyncFn from "@async-fn/jest";
import type { SendToViewArgs } from "../main/start-main-application/lens-window/application-window/lens-window-injection-token";
import sendToChannelInElectronBrowserWindowInjectable from "../main/start-main-application/lens-window/application-window/send-to-channel-in-electron-browser-window.injectable";
import { isEmpty } from "lodash/fp";
import enlistMessageChannelListenerInjectableInRenderer from "../renderer/channel/channel-listeners/enlist-message-channel-listener.injectable";
import enlistMessageChannelListenerInjectableInMain from "../main/channel/channel-listeners/enlist-message-channel-listener.injectable";
import enlistRequestChannelListenerInjectableInMain from "../main/channel/channel-listeners/enlist-request-channel-listener.injectable";
import sendToMainInjectable from "../renderer/channel/send-to-main.injectable";
import type { RequestChannel } from "../common/channel/request-channel-injection-token";

export const overrideIpcBridge = ({
  rendererDi,
  mainDi,
}: {
  rendererDi: DiContainer;
  mainDi: DiContainer;
}) => {
  const fakeRequestChannelMap = new Map<
    string,
    { promise: Promise<any>; resolve: (arg0: any) => Promise<void> }
  >();

  const requestChannelListenerFakesForMain = {
    set: <TChannel extends RequestChannel<unknown, unknown>>(
      channel: TChannel,
      callback: () => TChannel["_responseSignature"],
    ) => {
      const id = channel.id;

      if (!fakeRequestChannelMap.has(id)) {
        const mockInstance = asyncFn();

        fakeRequestChannelMap.set(id, {
          promise: mockInstance(),
          resolve: mockInstance.resolve,
        });
      }

      return fakeRequestChannelMap.get(id)?.resolve(callback);
    },

    get: <TChannel extends RequestChannel<unknown, unknown>>(channel: TChannel) => {
      const id = channel.id;

      if (!fakeRequestChannelMap.has(id)) {
        const mockInstance = asyncFn();

        fakeRequestChannelMap.set(id, {
          promise: mockInstance(),
          resolve: mockInstance.resolve,
        });
      }

      return fakeRequestChannelMap.get(id)?.promise;
    },
  };

  rendererDi.override(
    getValueFromChannelInjectable,

    () => async (channel) => {
      const callback = await requestChannelListenerFakesForMain.get(channel);

      return callback();
    },
  );

  mainDi.override(registerChannelInjectable, () => (channel, listener) => {
    requestChannelListenerFakesForMain.set(channel, listener);
  });

  const messageChannelListenerFakesForRenderer = new Map<
    string,
    ((...args: any[]) => void)[]
  >();

  mainDi.override(
    sendToChannelInElectronBrowserWindowInjectable,

    () =>
      (browserWindow, { channel: channelName, data = [] }: SendToViewArgs) => {
        const listeners = messageChannelListenerFakesForRenderer.get(channelName) || [];

        if (isEmpty(listeners)) {
          throw new Error(
            `Tried to send message to channel "${channelName}" but there where no listeners. Current channels with listeners: "${[
              ...messageChannelListenerFakesForRenderer.keys(),
            ].join('", "')}"`,
          );
        }

        listeners.forEach((handle) => handle(...data));
      },
  );

  const messageChannelListenerFakesForMain = new Map<
    string,
    ((...args: any[]) => void)[]
  >();

  const requestChannelListenerFakesForMain = new Map<
    string,
    ((...args: any[]) => any)[]
  >();

  rendererDi.override(
    enlistMessageChannelListenerInjectableInRenderer,

    () =>
      ({ channel, handler: newListener }) => {
        const oldListeners = messageChannelListenerFakesForRenderer.get(channel.id) || [];

        messageChannelListenerFakesForRenderer.set(channel.id, [...oldListeners, newListener]);

        return () => {};
      },
  );

  rendererDi.override(sendToMainInjectable, () => (channelId, message) => {
    const listeners = messageChannelListenerFakesForMain.get(channelId) || [];

    if (isEmpty(listeners)) {
      throw new Error(
        `Tried to send message to channel "${channelId}" but there where no listeners. Current channels with listeners: "${[
          ...messageChannelListenerFakesForMain.keys(),
        ].join('", "')}"`,
      );
    }

    listeners.forEach((handle) => handle(message));
  });

  mainDi.override(
    enlistMessageChannelListenerInjectableInMain,

    () =>
      ({ channel, handler: newListener }) => {
        const oldListeners = messageChannelListenerFakesForMain.get(channel.id) || [];

        messageChannelListenerFakesForMain.set(channel.id, [...oldListeners, newListener]);

        return () => {};
      },
  );

  mainDi.override(
    enlistRequestChannelListenerInjectableInMain,

    () =>
      ({ channel, handler }) => {
        // const existingHandles = mainIpcFakeHandles.get(channel.id) || [];

        // mainIpcFakeHandles.set(channel.id, [...existingHandles, handler]);

        return () => {};
      },
  );
};
