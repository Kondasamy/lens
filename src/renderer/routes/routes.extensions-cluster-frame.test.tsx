/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import type { DiContainer } from "@ogre-tools/injectable";
import { RenderResult, screen } from "@testing-library/react";
import { getDiForUnitTesting } from "../getDiForUnitTesting";
import React from "react";
import { computed, runInAction } from "mobx";
import rendererExtensionsInjectable from "../../extensions/renderer-extensions.injectable";
import { LensRendererExtension } from "../../extensions/lens-renderer-extension";
import observableHistoryInjectable from "../navigation/observable-history.injectable";
import { createMemoryHistory, MemoryHistory } from "history";
import { createObservableHistory } from "mobx-observable-history";
import currentlyInClusterFrameInjectable from "./currently-in-cluster-frame.injectable";
import directoryForUserDataInjectable from "../../common/app-paths/directory-for-user-data/directory-for-user-data.injectable";

import type {
  ClusterPageMenuRegistration,
  PageRegistration,
} from "../../extensions/registries";
import { renderFor } from "../components/test-utils/renderFor";
import { Observer } from "mobx-react";
import type { Cluster } from "../../common/cluster/cluster";
import hostedClusterInjectable from "../../common/cluster-store/hosted-cluster.injectable";
import customResourceDefinitionsInjectable from "../components/+custom-resources/custom-resources.injectable";
import type { CustomResourceDefinition } from "../../common/k8s-api/endpoints";
import currentRouteComponentInjectable from "./current-route-component.injectable";

describe("routes - extensions - cluster frame", () => {
  let history: MemoryHistory;
  let di: DiContainer;
  let rendered: RenderResult;

  beforeEach(async () => {
    di = getDiForUnitTesting({ doGeneralOverrides: true });

    di.override(
      directoryForUserDataInjectable,
      () => "some-directory-for-user-data",
    );

    const extension = new TestExtension({
      id: "@some/extension-id",

      clusterPages: [
        {
          components: {
            Page: () => <div data-testid="some-route-from-extension" />,
          },
        },
        {
          id: "//some-child-page/",
          components: {
            Page: () => <div data-testid="some-child-page-from-extension"/>,
          },
        },
        {
          id: "some-other-child-page",
          components: {
            Page: () => <div data-testid="some-other-child-page-from-extension" />,
          },
        },
      ],

      clusterPageMenus: [
        {
          id: "some-parent-id",
          title: "Parent",
          components: {
            Icon: () => <div>Some icon</div>,
          },
        },
        {
          id: "some-child-id",
          parentId: "some-parent-id",
          target: { pageId: "//some-child-page/" },
          title: "Child 1",

          components: {
            Icon: () => <div>Some icon</div>,
          },
        },
        {
          id: "some-other-child",
          parentId: "some-parent-id",
          target: { pageId: "some-other-child-page" },
          title: "Child 2",
          components: {
            Icon: () => <div>Some icon</div>,
          },
        },
      ],

      globalPages: [],
    });

    di.override(rendererExtensionsInjectable, () =>
      computed((): LensRendererExtension[] => [extension]),
    );

    history = createMemoryHistory();
    const observableHistory = createObservableHistory(history);

    di.override(observableHistoryInjectable, () => observableHistory);
    di.override(currentlyInClusterFrameInjectable, () => true);

    const clusterStub = {
      allowedResources: [],
    } as Cluster;

    di.override(hostedClusterInjectable, () => clusterStub);

    di.override(customResourceDefinitionsInjectable, () =>
      computed((): CustomResourceDefinition[] => []),
    );

    await di.runSetups();

    const render = renderFor(di);

    const currentRouteComponent = di.inject(currentRouteComponentInjectable);

    rendered = render(
      <Observer>
        {() => {
          const Component = currentRouteComponent.get();

          if (!Component) {
            return <div data-testid="no-active-route" />;
          }

          return <Component />;
        }}
      </Observer>,
    );
  });

  it("renders without active route", () => {
    screen.getByTestId("no-active-route");
  });

  describe("when navigating to extension front page", () => {
    beforeEach(() => {
      runInAction(() => {
        history.replace("/extension/some--extension-id");
      });
    });

    it("renders", () => {
      expect(rendered.container).toMatchSnapshot();
    });

    it("shows component for parent route", () => {
      screen.getByTestId("some-route-from-extension");
    });

    describe("when navigating to extension child page", () => {
      beforeEach(() => {
        runInAction(() => {
          history.replace("/extension/some--extension-id/some-child-page");
        });
      });

      it("renders", () => {
        expect(rendered.container).toMatchSnapshot();
      });

      it("shows component for child route having matching ID", () => {
        screen.getByTestId("some-child-page-from-extension");
      });
    });
  });
});

class TestExtension extends LensRendererExtension {
  constructor({
    id,
    globalPages,
    clusterPages,
    clusterPageMenus,
  }: {
    id: string;
    globalPages: PageRegistration[];
    clusterPages: PageRegistration[];
    clusterPageMenus: ClusterPageMenuRegistration[];
  }) {
    super({
      id,
      absolutePath: "irrelevant",
      isBundled: false,
      isCompatible: false,
      isEnabled: false,
      manifest: { name: id, version: "some-version" },
      manifestPath: "irrelevant",
    });

    this.globalPages = globalPages;
    this.clusterPages = clusterPages;
    this.clusterPageMenus = clusterPageMenus;
  }
}
