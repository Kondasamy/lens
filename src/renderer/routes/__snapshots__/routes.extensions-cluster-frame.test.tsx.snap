// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`routes - extensions - cluster frame when navigating to extension front page renders 1`] = `
<div>
  <div
    data-testid="some-route-from-extension"
  />
</div>
`;

exports[`routes - extensions - cluster frame when navigating to extension front page when navigating to extension child page renders 1`] = `
<div>
  <div
    class="TabLayout"
  >
    <div
      class="Tabs center scrollable"
    >
      <div
        class="Tab flex gaps align-center active"
        tabindex="0"
      >
        <div
          class="label"
        >
          Child 1
        </div>
      </div>
      <div
        class="Tab flex gaps align-center"
        tabindex="0"
      >
        <div
          class="label"
        >
          Child 2
        </div>
      </div>
    </div>
    <main
      class=""
    >
      <div
        data-testid="some-child-page-from-extension"
      />
    </main>
  </div>
</div>
`;
